docs/secure-communication-agents.asciidoc

```diff
@@ -1,21 +1,353 @@
+
+The following example invalidates the `java-001` API key.
+
+["source","sh",subs="attributes"]
+-----
+{beatname_lc} apikey invalidate --name java-001
+-----
+
+The response will look similar to this:
+
+[source,console-result,subs="attributes,callouts"]
+--------------------------------------------------
+Invalidated keys ... qT4tz28B1g59zC3uAXfW
+Error count ........ 0
+--------------------------------------------------
+
+A full list of `apikey` subcommands and flags is available in the <<apikey-command,API key command reference>>.
+
+[[set-api-key]]
+[float]
+=== Set the API key in your APM Agents
+
+You can now apply your newly created API keys in the configuration of each of your APM Agents.
+See the relevant Agent documentation for additional information:
+
+// Agent meta: https://github.com/elastic/apm/issues/183
+
+// GOOD DOCS:
+* *Go Agent*: {apm-go-ref}/configuration.html#config-api-key[`ELASTIC_APM_API_KEY`]
+// MERGED: https://github.com/elastic/apm-agent-go/pull/698
+
+// No issue or docs yet
+// * *Java Agent*: {apm-java-ref}/config-reporter.html#config-api-key[`api_key`]
+
+// No issue or docs yet
+// * *.NET Agent*: {apm-dotnet-ref}/config-reporter.html#[`API_KEY`]
+
+// No issue or docs yet
+// * *Node.js Agent*: {apm-node-ref}/configuration.html[`api_key`]
+
+// No docs yet
+// * *Python Agent*: {apm-py-ref}/configuration.html#config-api-key[`api_key`]
+// WIP
+
+// No docs yet
+// * *Ruby Agent*: {apm-ruby-ref}/configuration.html#config-api-key[`api_key`]
+// MERGED: https://github.com/elastic/apm-agent-ruby/pull/655
+
+[[api-key-settings]]
+=== API key configuration options
+
+You can specify the following options in the `apm-server.api_key.*` section of the
++{beatname_lc}.yml+ configuration file.
+They apply to API key communication between the APM Server and APM Agents.
+
+// To do!
+// These will become links, but the relevant docs haven't been merged yet.
+These are different from the API key settings used for the Elasticsearch output and monitoring.
+
+[float]
+===== `enabled`
+
+Enable API key authorization by setting `enabled` to `true`.
+Agents will include a valid API key in the following format: `Authorization: ApiKey <token>`.
+The key must be the base64 encoded representation of the API key's `id:key`.
+By default, `enabled` is set to `false`, and API key support is disabled.
+
+[float]
+===== `limit`
+
+Each unique API key triggers one request to Elasticsearch.
+This setting restricts the number of unique API keys are allowed per minute.
+The minimum value for this setting should be the number of API keys configured in your monitored services.
+The default `limit` is `100`.
 
-[[secret-token]]
 [float]
+===== `elasticsearch.hosts`
+
+API keys are fetched from Elasticsearch.
+This configuration needs to point to a secured Elasticsearch cluster that is able to serve API key requests.
+If nothing is configured, configuration settings from the `output` section will be reused.
```

Comment:
Can you mention this once for the whole `elasticsearch.*` section. The way it is described now one might think that settings from `apm-server.api_key.elasticsearch` and `output.elasticsearch` are merged, but they are not. If some `apm-server.api_key.elasticsearch.` setting is specificed, then `output.elasticsearch` is ignored here.